File: generate_gif.py
import os
import math
from PIL import Image, ImageDraw, ImageFont

def create_gif():
    frames = []
    width, height = 800, 500
    num_frames = 90
    
    # Define people properties: (x_center_ratio, y_center_ratio, size_ratio, speed)
    people = [
        {"x_start": 0.25, "y_start": 0.45, "x_end": 0.15, "y_end": 0.65, "size_start": 20, "size_end": 70, "name": "EMP-8821", "status": "AUTHORIZED", "color": "#00FF66"},
        {"x_start": 0.5, "y_start": 0.50, "x_end": 0.5, "y_end": 0.75, "size_start": 25, "size_end": 85, "name": "EMP-1093", "status": "AUTHORIZED", "color": "#00FF66"},
        {"x_start": 0.75, "y_start": 0.40, "x_end": 0.85, "y_end": 0.60, "size_start": 18, "size_end": 55, "name": "UNKNOWN", "status": "UNAUTHORIZED", "color": "#FF3333"},
        {"x_start": 0.38, "y_start": 0.42, "x_end": 0.35, "y_end": 0.55, "size_start": 15, "size_end": 45, "name": "VIS-5521", "status": "VISITOR", "color": "#00CCFF"},
    ]

    try:
        font_large = ImageFont.truetype("C:\\Windows\\Fonts\\consola.ttf", 16)
        font_small = ImageFont.truetype("C:\\Windows\\Fonts\\consola.ttf", 12)
    except:
        font_large = ImageFont.load_default()
        font_small = ImageFont.load_default()

    for i in range(num_frames):
        # Create a dark tech background
        img = Image.new('RGB', (width, height), color=(15, 20, 30))
        draw = ImageDraw.Draw(img)
        
        # Draw some tech grid lines
        for x in range(0, width, 40):
            draw.line([(x, 0), (x, height)], fill=(25, 35, 50), width=1)
        for y in range(0, height, 40):
            draw.line([(0, y), (width, y)], fill=(25, 35, 50), width=1)
            
        progress = i / (num_frames - 1)
        
        # Sort people by size so larger (closer) ones are drawn last
        current_people_state = []
        for person in people:
            x = width * (person["x_start"] + (person["x_end"] - person["x_start"]) * progress)
            y = height * (person["y_start"] + (person["y_end"] - person["y_start"]) * progress)
            size = person["size_start"] + (person["size_end"] - person["size_start"]) * progress
            current_people_state.append((size, x, y, person))
            
        current_people_state.sort(key=lambda item: item[0])
        
        for size, x, y, person in current_people_state:
            
            # Draw person (simple silhouette: circle for head, rounded rect for body)
            head_radius = size
            head_bbox = [x - head_radius, y - head_radius*2, x + head_radius, y]
            draw.ellipse(head_bbox, fill=(80, 100, 130))
            
            body_bbox = [x - head_radius*1.5, y + 5, x + head_radius*1.5, y + head_radius*4]
            # using rectangle for modern look
            draw.rectangle(body_bbox, fill=(60, 80, 110))
            
            # Face detection logic
            # Offset it based on person for staggered detection
            identification_threshold = 0.3 + (people.index(person) * 0.1)
            
            if progress > identification_threshold - 0.2:
                # bounding box
                box_pad = size * 0.2
                bbox = [head_bbox[0] - box_pad, head_bbox[1] - box_pad, head_bbox[2] + box_pad, head_bbox[3] + box_pad]
                
                # Draw box only if fully identified, otherwise draw scanning lines
                if progress > identification_threshold:
                    draw.rectangle(bbox, outline=person["color"], width=max(1, int(size/15)))
                    
                    # Corner accents
                    cl = size * 0.4 # corner length
                    cw = max(2, int(size/10))
                    # top left
                    draw.line([(bbox[0], bbox[1]+cl), (bbox[0], bbox[1]), (bbox[0]+cl, bbox[1])], fill=person["color"], width=cw)
                    # top right
                    draw.line([(bbox[2]-cl, bbox[1]), (bbox[2], bbox[1]), (bbox[2], bbox[1]+cl)], fill=person["color"], width=cw)
                    # bottom left
                    draw.line([(bbox[0], bbox[3]-cl), (bbox[0], bbox[3]), (bbox[0]+cl, bbox[3])], fill=person["color"], width=cw)
                    # bottom right
                    draw.line([(bbox[2]-cl, bbox[3]), (bbox[2], bbox[3]), (bbox[2], bbox[3]-cl)], fill=person["color"], width=cw)
    
                else:
                    # Scanning effect
                    scan_y = bbox[1] + (bbox[3] - bbox[1]) * ((progress - (identification_threshold - 0.2)) / 0.2)
                    draw.line([(bbox[0], scan_y), (bbox[2], scan_y)], fill=(0, 255, 255), width=2)
                
            # Identify after threshold
            if progress > identification_threshold:
                # popup line
                line_end_x = bbox[2] + size * 0.5
                line_end_y = bbox[1] - size * 0.5
                draw.line([(bbox[2], bbox[1]), (line_end_x, line_end_y)], fill=person["color"], width=2)
                
                # popup box
                popup_w, popup_h = 120, 50
                popup_bbox = [line_end_x, line_end_y - popup_h, line_end_x + popup_w, line_end_y]
                
                # Semi-transparent background for popup
                # PIL doesn't support drawing transparent rect directly on RGB without alpha compositing
                # We'll just draw a solid very dark rect
                draw.rectangle(popup_bbox, fill=(20, 25, 35), outline=person["color"], width=1)
                
                # Text
                text_y = line_end_y - popup_h + 8
                draw.text((line_end_x + 8, text_y), person["name"], font=font_large, fill=person["color"])
                draw.text((line_end_x + 8, text_y + 22), person["status"], font=font_small, fill=person["color"])

        # HUD overlay elements
        draw.text((20, 20), "SYSTEM: ACTIVE", font=font_large, fill=(0, 255, 255))
        draw.text((20, 45), f"TRACKING: {len(current_people_state)} SUBJECTS", font=font_small, fill=(0, 255, 255))
        draw.text((width - 150, 20), "REC: 1080p 60FPS", font=font_large, fill=(255, 50, 50))
        
        frames.append(img)
        
    output_path = r"d:\Swarag\Workspace\Qlsdynamics\qlsdynamics-website\public\images\dynamic_face_tracking.gif"
    frames[0].save(output_path, save_all=True, append_images=frames[1:], optimize=False, duration=50, loop=0)
    print(f"GIF built at {output_path}")

create_gif()

File: generate_gif.js
const PImage = require('pureimage');
const fs = require('fs');
const GIFEncoder = require('gif-encoder-2');

async function createGif() {
    const width = 800;
    const height = 500;
    const numFrames = 90;

    const encoder = new GIFEncoder(width, height);
    encoder.start();
    encoder.setRepeat(0); // 0 for repeat, -1 for no-repeat
    encoder.setDelay(50); // frame delay in ms
    encoder.setQuality(10); // image quality. 10 is default.

    // Define people properties: (x_start_ratio, y_start_ratio, x_end_ratio, y_end_ratio, size_start, size_end, name, status, color)
    const people = [
        { x_start: 0.15, y_start: 0.45, x_end: 0.15, y_end: 0.65, size_start: 22, size_end: 75, status: "AUTHORIZED", color: "#00FF66" },
        { x_start: 0.38, y_start: 0.40, x_end: 0.38, y_end: 0.60, size_start: 18, size_end: 60, status: "VISITOR", color: "#00CCFF" },
        { x_start: 0.62, y_start: 0.48, x_end: 0.62, y_end: 0.70, size_start: 24, size_end: 80, status: "AUTHORIZED", color: "#00FF66" },
        { x_start: 0.85, y_start: 0.42, x_end: 0.85, y_end: 0.62, size_start: 20, size_end: 70, status: "UNAUTHORIZED", color: "#FF3333" },
    ];

    const fnt = PImage.registerFont('C:\\Windows\\Fonts\\consola.ttf', 'Consolas');
    await new Promise((resolve, reject) => {
        try {
            fnt.loadSync();
            resolve();
        } catch (e) {
            resolve();
        }
    });

    for (let i = 0; i < numFrames; i++) {
        const img = PImage.make(width, height);
        const ctx = img.getContext('2d');

        ctx.fillStyle = "#0F141E";
        ctx.fillRect(0, 0, width, height);

        ctx.strokeStyle = "#192332";
        ctx.lineWidth = 1;
        for (let x = 0; x < width; x += 40) {
            ctx.beginPath(); ctx.moveTo(x, 0); ctx.lineTo(x, height); ctx.stroke();
        }
        for (let y = 0; y < height; y += 40) {
            ctx.beginPath(); ctx.moveTo(0, y); ctx.lineTo(width, y); ctx.stroke();
        }

        const progress = i / (numFrames - 1);

        let current_people_state = people.map((p, idx) => {
            const x = width * (p.x_start + (p.x_end - p.x_start) * progress);
            const y = height * (p.y_start + (p.y_end - p.y_start) * progress);
            const size = p.size_start + (p.size_end - p.size_start) * progress;
            return { size, x, y, person: p, idx };
        });

        current_people_state.sort((a, b) => a.size - b.size);

        for (const { size, x, y, person, idx } of current_people_state) {

            ctx.fillStyle = "#506482";
            ctx.beginPath();
            ctx.arc(x, y - size, size, 0, Math.PI * 2);
            ctx.fill();

            ctx.fillStyle = "#3C506E";
            ctx.beginPath();
            const bw = size * 1.5;
            const bh = size * 4;
            ctx.fillRect(x - bw, y + 5, bw * 2, bh);

            const identification_threshold = 0.3 + (idx * 0.1);

            if (progress > identification_threshold - 0.2) {
                const box_pad = size * 0.2;
                const bx = x - size - box_pad;
                const by = y - size * 2 - box_pad;
                const bw2 = (size + box_pad) * 2;
                const bh2 = (size * 2 + box_pad);

                if (progress > identification_threshold) {
                    ctx.strokeStyle = person.color;
                    ctx.lineWidth = Math.max(1, Math.floor(size / 15));
                    ctx.strokeRect(bx, by, bw2, bh2);

                    const cl = size * 0.4;
                    const cw = Math.max(2, Math.floor(size / 10));
                    ctx.strokeStyle = person.color;
                    ctx.lineWidth = cw;

                    ctx.beginPath(); ctx.moveTo(bx, by + cl); ctx.lineTo(bx, by); ctx.lineTo(bx + cl, by); ctx.stroke();
                    ctx.beginPath(); ctx.moveTo(bx + bw2 - cl, by); ctx.lineTo(bx + bw2, by); ctx.lineTo(bx + bw2, by + cl); ctx.stroke();
                    ctx.beginPath(); ctx.moveTo(bx, by + bh2 - cl); ctx.lineTo(bx, by + bh2); ctx.lineTo(bx + cl, by + bh2); ctx.stroke();
                    ctx.beginPath(); ctx.moveTo(bx + bw2 - cl, by + bh2); ctx.lineTo(bx + bw2, by + bh2); ctx.lineTo(bx + bw2, by + bh2 - cl); ctx.stroke();

                    const p_start_x = bx + bw2 / 2;
                    const p_start_y = by;
                    let p_end_x = bx + bw2 / 2;
                    let p_end_y = by - 30;

                    // Fan out the popups securely so they don't overlap vertically/horizontally
                    if (x < width * 0.25) {
                        p_end_x -= 100;
                        p_end_y -= 10;
                    } else if (x < width * 0.50) {
                        p_end_x -= 40;
                        p_end_y -= 50;
                    } else if (x < width * 0.75) {
                        p_end_x += 40;
                        p_end_y -= 10;
                    } else {
                        p_end_x += 100;
                        p_end_y -= 50;
                    }

                    const pw = 130;
                    const ph = 30;

                    let px = p_end_x - pw / 2;
                    // Cap horizontal position
                    if (px < 5) px = 5;
                    if (px + pw > width - 5) px = width - pw - 5;

                    const py = p_end_y - ph;

                    ctx.strokeStyle = person.color;
                    ctx.lineWidth = 2;
                    ctx.beginPath(); ctx.moveTo(p_start_x, p_start_y); ctx.lineTo(p_end_x, p_end_y); ctx.stroke();

                    ctx.fillStyle = "#141923";
                    ctx.fillRect(px, py, pw, ph);
                    ctx.strokeStyle = person.color;
                    ctx.lineWidth = 1;
                    ctx.strokeRect(px, py, pw, ph);


                    ctx.fillStyle = person.color;
                    ctx.font = "12pt Consolas";
                    ctx.fillText(person.status, px + 8, py + 20);

                } else {
                    const scan_progress = (progress - (identification_threshold - 0.2)) / 0.2;
                    const scan_y = by + bh2 * scan_progress;
                    ctx.strokeStyle = "#00FFFF";
                    ctx.lineWidth = 2;
                    ctx.beginPath(); ctx.moveTo(bx, scan_y); ctx.lineTo(bx + bw2, scan_y); ctx.stroke();
                }
            }
        }

        ctx.fillStyle = "#00FFFF";
        ctx.font = "16pt Consolas";
        ctx.fillText("SYSTEM: ACTIVE", 20, 30);

        encoder.addFrame(ctx);
        if (i % 10 === 0) console.log("Rendered frame", i);
    }

    encoder.finish();
    const buffer = encoder.out.getData();
    fs.writeFileSync('./public/images/dynamic_face_tracking.gif', buffer);
    console.log("GIF successfully generated! Size:", buffer.length);
}

createGif().catch(console.error);
